c-dependencies/js-compute-runtime/builtins/backend.cpp

```diff
@@ -1,22 +1,605 @@
+
+    auto it =
+        std::stable_partition(ciphers.begin(), ciphers.end(), byKeyExchange(KeyExchange::EECDH));
+
+    /* AES is our preferred symmetric cipher */
+    auto aes = {Encryption::AES128, Encryption::AES128GCM, Encryption::AES256,
+                Encryption::AES256GCM};
+
+    /* Now arrange all ciphers by preference: */
+    it = std::stable_partition(it, ciphers.end(), byEncryption(aes));
+
+    /* Move ciphers without forward secrecy to the end */;
+    std::stable_partition(it, ciphers.end(), [compare = byKeyExchange(KeyExchange::RSA)](auto &c) {
+      return !compare(c);
+    });
+  }
+
+  std::function<bool(const Cipher &, const Cipher &)> compareKeyExchange(KeyExchange val) {
+    return [val](auto &left, auto &right) {
+      if (left.kx == val) {
+        return right.kx != val;
+      } else {
+        return right.kx == val;
+      }
+    };
+  }
+
+  std::function<bool(const Cipher &)> byProtocol(Protocol val) {
+    return [val](auto &c) { return c.protocol == val; };
+  }
+
+  std::function<bool(const Cipher &)> byKeyExchange(KeyExchange val) {
+    return [val](auto &c) { return c.kx == val; };
+  }
+
+  std::function<bool(const Cipher &)> byAuthentication(Authentication val) {
+    return [val](auto &c) { return c.au == val; };
+  }
+
+  std::function<bool(const Cipher &)> byEncryption(std::set<Encryption> vals) {
+    return [vals](auto &c) { return vals.find(c.enc) != vals.end(); };
+  }
+
+  std::function<bool(const Cipher &)> byEncryption(Encryption val) {
+    return [val](auto &c) { return c.enc == val; };
+  }
+
+  std::function<bool(const Cipher &)> byEncryptionLevel(EncryptionLevel val) {
+    return [val](auto &c) { return c.level == val; };
+  }
+
+  std::function<bool(const Cipher &)> byMessageDigest(MessageDigest val) {
+    return [val](auto &c) { return c.mac == val; };
+  }
+
+  std::vector<std::string_view> split(std::string_view s, std::string_view delimiter) {
+    size_t pos_start = 0, pos_end, delim_len = delimiter.length();
+    std::string token;
+    std::vector<std::string> res;
+
+    while ((pos_end = s.find(delimiter, pos_start)) != std::string::npos) {
+      token = s.substr(pos_start, pos_end - pos_start);
+      pos_start = pos_end + delim_len;
+      res.push_back(token);
+    }
+
+    res.push_back(s.substr(pos_start));
+    return res;
+  }
+
+  std::vector<std::string_view> splitCipherSuiteString(std::string_view string) {
+    std::vector<std::string> result;
+    std::stringstream stringStream(string);
+    std::string line;
+    while (std::getline(stringStream, line, ':')) {
+      std::size_t prev = 0, pos;
+      while ((pos = line.find_first_of(" ,", prev)) != std::string::npos) {
+        if (pos > prev) {
+          result.push_back(line.substr(prev, pos - prev));
+        }
+        prev = pos + 1;
+      }
+      if (prev < line.length()) {
+        result.push_back(line.substr(prev, std::string::npos));
+      }
+    }
+    return result;
+  }
```

Comment:
How about refactoring this loop like the following to avoid needing to use the `std::stringstream`?
https://godbolt.org/z/T9qhGM7YG

Reply:
I like it 😊